src/proc_fuse.c

```diff
@@ -473,6 +493,147 @@ static void get_blkio_io_value(char *str, unsigned major, unsigned minor,
 	}
 }
 
+struct devinfo {
+	char *name;
+	unsigned int major;
+	unsigned int minor;
+	struct devinfo *next;
+};
+
+int getns(pid_t pid, const char *ns_type)
+{
+	char fpath[100];
+	memset(fpath, 0, sizeof(fpath));
+	snprintf(fpath, 99, "/proc/%d/ns/%s", pid, ns_type);
+	return open(fpath, O_RDONLY);
+}
+
+struct devinfo* container_dev_read(pid_t pid) {
+	int nsfd;
+	DIR *dir;
+	struct dirent *ptr;
+	struct stat dev_stat;
+	struct devinfo *head = NULL, *end;
+	char fpath[261], dev_name[256];
+	pid_t child_pid;
+	int mypipe[2];
+	int dev_num;
+	FILE *stream;
+
+	if (pipe(mypipe) < 0) {
+		perror("Error creating pipe");
+		return head;
+	}
+
+	child_pid = fork();
+	if (child_pid < 0) {
+		close(mypipe[0]);
+		close(mypipe[1]);
+		perror("Error forking child");
+		return head;
+	}
+	if (child_pid == 0) {
+		/* Disallow signal reception in child process */
+		sigset_t oldset;
+		sigset_t newset;
+		sigemptyset(&newset);
+		sigaddset(&newset, SIGTERM);
+		sigaddset(&newset, SIGINT);
+		sigaddset(&newset, SIGHUP);
+		sigaddset(&newset, SIGQUIT);
+		sigprocmask(SIG_BLOCK,&newset,&oldset);
+
+		close(mypipe[0]);
+		stream = fdopen(mypipe[1], "w");
+		if (stream == NULL) {
+			lxcfs_error("Error opening pipe for writing: %s\n", strerror(errno));
+			goto child_out;
+		}
+		nsfd = getns(pid, "mnt");
+		if (nsfd < 0) {
+			lxcfs_error("Error getting mnt ns: %s\n", strerror(errno));
+			goto child_out;
+		}
+		if (setns(nsfd, 0) < 0) {
+			lxcfs_error("Error setting mnt ns: %s\n", strerror(errno));
+			goto child_out;
+		}
+		dir = opendir("/dev");
+		if (dir == NULL) {
+			lxcfs_error("Error opening dir /dev: %s\n", strerror(errno));
+			goto child_out;
+		}
+		while ((ptr = readdir(dir)) != NULL) {
```

Comment:
here you are iterating over the `/dev` inside the container mount namespace to make a list of block devices which are used by the container, right?

What do you think about using `/proc/<pid>/mountinfo` file instead to get all used block devices? Speaking more precisely, this file allows you to get all block devices which are *mounted* inside the container mount namespace.

Another limitation here is that in the container there can be a few different mount namespaces.

Reply:
Good idea! Thank you for your suggestion.

For the first question, I did this by going into the container's namespace and traversing the `/dev` directory underneath it.

I had thought about **using `/proc/PID/mountinfo` on the host side** (If I understand correctly) instead of going into the container namespace to achieve this functionality. But I have encountered some problems. 

1.  I cannot get the unmounted block device from this file; in fact, the block device mounted through `--device` cannot be obtained through the above method.
```bash
[root@localhost ~]# docker run -tid --name test --device /dev/sdb:/dev/xxx  rnd-dockerhub.huawei.com/official/ubuntu-arm64 bash
c82db9506ffaff30ce480852d3133cd4528a013d427372c77d91621edfb75809
[root@localhost ~]# docker inspect --format '{{.State.Pid}}'  c82db9506ffaff30ce480852d313
3325697
[root@localhost ~]# cat /proc/3325697/mountinfo | grep sdb
[root@localhost ~]# cat /proc/3325697/mountinfo | grep xxx
[root@localhost ~]# cat /proc/3325697/mountinfo | grep /dev
520 518 0:59 / /dev rw,nosuid - tmpfs tmpfs rw,seclabel,size=65536k,mode=755
521 520 0:60 / /dev/pts rw,nosuid,noexec,relatime - devpts devpts rw,seclabel,gid=5,mode=620,ptmxmode=666
530 523 0:37 /docker/c82db9506ffaff30ce480852d3133cd4528a013d427372c77d91621edfb75809 /sys/fs/cgroup/devices ro,nosuid,nodev,noexec,relatime master:11 - cgroup cgroup rw,seclabel,devices
540 520 0:57 / /dev/mqueue rw,nosuid,nodev,noexec,relatime - mqueue mqueue rw,seclabel
541 518 253:0 /var/lib/docker/containers/c82db9506ffaff30ce480852d3133cd4528a013d427372c77d91621edfb75809/resolv.conf /etc/resolv.conf rw,relatime - ext4 /dev/mapper/euleros-root rw,seclabel
542 518 253:0 /var/lib/docker/containers/c82db9506ffaff30ce480852d3133cd4528a013d427372c77d91621edfb75809/hostname /etc/hostname rw,relatime - ext4 /dev/mapper/euleros-root rw,seclabel
543 518 253:0 /var/lib/docker/containers/c82db9506ffaff30ce480852d3133cd4528a013d427372c77d91621edfb75809/hosts /etc/hosts rw,relatime - ext4 /dev/mapper/euleros-root rw,seclabel
544 520 0:56 / /dev/shm rw,nosuid,nodev,noexec,relatime - tmpfs shm rw,seclabel,size=65536k
464 520 0:60 /0 /dev/console rw,nosuid,noexec,relatime - devpts devpts rw,seclabel,gid=5,mode=620,ptmxmode=666
```

As you can see, I didn't get this device. However, I can get it in the following way:
```bash
[root@localhost ~]# docker exec -it test bash -c "ls /dev/xxx"
/dev/xxx
```

2. An important reason why I enter the `/dev` device to read the device name is that the device names in the container and the host are different.

Hope I understand you correctly. Welcome to exchange your ideas

Reply:
>I cannot get the unmounted block device from this file; in fact, the block device mounted through --device cannot be obtained through the above method.

yes, you can only get devices which are *mounted* inside the container using this method. `--device` option in Docker does *not* mount device, but just (I guess) performs a bind-mount of it inside the container.